specs/APIReview_TextureStream.md

```diff
@@ -0,0 +1,486 @@
+    return S_OK;
+  }).Get(), &post_stop_token);
+
+HRESULT CallWebView2API(bool createBuffer,
+                        UINT32 width,
+                        UINT32 height,
+                        bool requestAvailableBufer,
+                        bool sendTexture,
+                        ICoreWebView2StagingTexture* send_texture,
+                        UINT64 timestamp,
+                        ICoreWebView2StagingTexture** texture) {
+  if (createBuffer) {
+    // Create shared buffer.
+    return webviewTextureStream->CreateSharedBuffer(width, height, &texture);
+  }
+
+  if (requestAvailableBufer) {
+    return webviewTextureStream->GetAvailableBuffer(&texture);
+  }
+
+  if (sendTexture) {
+    // Notify the renderer for updated texture on the shared buffer.
+    webviewTextureStream->SetBuffer(send_texture, timestamp);
+    webviewTextureStream->Present();
+  }
+}
+```
+
+# API Details
+```
+[v1_enum]
+typedef enum COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND {
+  /// The host can't create a TextureStream instance more than once
+  /// for a specific stream id.
+  COREWEBVIEW2_TEXTURE_STREAM_ERROR_STREAM_ID_ALREADY_REGISTERED,
+  /// Occurs when the host calls CreateBuffer or Present
+  /// APIs without being called of Start event. Or, 10 seconds passed before
+  /// calling these APIs since the OnStart event.
+  COREWEBVIEW2_TEXTURE_STREAM_ERROR_NO_VIDEO_TRACK_STARTED,
+  /// The buffer has been removed using RemoveBuffer.
+  COREWEBVIEW2_TEXTURE_STREAM_ERROR_BUFFER_NOT_FOUND,
+  /// The texture to be presented is already in use for rendering.
+  /// Call GetAvailableBuffer to determine an available buffer to present.
+  /// The developer can technically call SetBuffer multiple times.
+  /// But once they call Present, the buffer becomes "in use" until
+  /// they call SetBuffer and Present on a different buffer and wait a bit
+  /// for the original buffer to stop being used.
+  COREWEBVIEW2_TEXTURE_STREAM_ERROR_BUFFER_IN_USE,
+} COREWEBVIEW2_TEXTURE_STREAM_ERROR_KIND;
+/// This is ICoreWebView2Staging3 that returns the texture stream interface.
+[uuid(96c27a45-f142-4873-80ad-9d0cd899b2b9), object, pointer_default(unique)]
+interface ICoreWebView2Staging3 : IUnknown {
+  /// Registers the stream id that the host can handle, providing a
+  /// texture stream when requested from the WebView2's JavaScript code.
+  /// The host can register multiple unique stream instances, each with
+  /// a unique stream ID, enabling the host to stream from different sources
+  /// concurrently.
+  /// The host should call this only once for unique streamId. The second
+  /// call of already created streamId without destroying
+  /// ICoreWebView2StagingTextureStream object will return an error.
+  /// 'd3dDevice' is used for creating shared IDXGI resource and NT shared
+  /// of it. The host should use Adapter of the LUID from the GetRenderAdapterLUID
+  /// for creating the D3D Device.
+  HRESULT CreateTextureStream(

```

Comment:
- TextureStream object should be video source, which is like single camera in the Chromium engine perspective. So, it isn't doable for single TextureStream object to have multiple names; The name is like device name. When the host supports multiple names, then it has to create multiple TextureStream that will be clear even with the event, otherwise every events 

- CreateTextureStream needs D3DDevice object, which can be created from LUID that the host can get from GetRenderAdapterLUID.   GetRenderAdapterLUID API can successfully return a value only when GPU process is created that usually happens in navigation.  

- TextureStream is associated with its counterpart in the Javascript, getTextureStream and registerTextureStream. These Javascript APIs can be called only in the page (window) so it looks more natural for it to be in CoreWindow2.
-

Reply:
From discussion, by the scenario we need one texturestream to be able to work with multiple CoreWebView2s. Since the texturestream isn't tied to one CoreWebView2 we should move these methods to the CoreWebView2Environment. On the environment you should be able to CreateTextureStream. But as you pointed out there is no connection required to particular CoreWebView2 objects, so there's no Add or Register method necessary. Just create the TextureStream with a particular name and then its available by that name in all the CoreWebView2s for that environment.

We didn't talk as much about the RenderAdapterLUID property or the RenderAdapterLUIDChanged event, but it sounds like maybe those also should move to the CoreWebView2Environment.